File: backend/services/context-gateway-go/internal/api/federation.go
// Federation GraphQL endpoint for Apollo Federation integration
package api

import (
	"context"
	"encoding/json"
	"net/http"
	"log"

	"github.com/gorilla/mux"
	pb "context-gateway-go/proto"
)

// FederationHandler handles GraphQL federation requests
type FederationHandler struct {
	contextService ContextGatewayService
}

// ContextGatewayService interface for federation
type ContextGatewayService interface {
	GetSnapshot(ctx context.Context, req *pb.GetSnapshotRequest) (*pb.ClinicalSnapshot, error)
	LoadRecipe(ctx context.Context, req *pb.LoadRecipeRequest) (*pb.WorkflowRecipe, error)
	GetServiceHealth(ctx context.Context, req *pb.HealthRequest) (*pb.HealthResponse, error)
	GetMetrics(ctx context.Context, req *pb.MetricsRequest) (*pb.MetricsResponse, error)
}

// NewFederationHandler creates a new federation handler
func NewFederationHandler(contextService ContextGatewayService) *FederationHandler {
	return &FederationHandler{
		contextService: contextService,
	}
}

// GraphQLRequest represents a GraphQL request
type GraphQLRequest struct {
	Query         string                 `json:"query"`
	Variables     map[string]interface{} `json:"variables,omitempty"`
	OperationName string                 `json:"operationName,omitempty"`
}

// GraphQLResponse represents a GraphQL response
type GraphQLResponse struct {
	Data   interface{} `json:"data,omitempty"`
	Errors []GraphQLError `json:"errors,omitempty"`
}

// GraphQLError represents a GraphQL error
type GraphQLError struct {
	Message    string                 `json:"message"`
	Locations  []GraphQLLocation      `json:"locations,omitempty"`
	Path       []interface{}          `json:"path,omitempty"`
	Extensions map[string]interface{} `json:"extensions,omitempty"`
}

// GraphQLLocation represents a location in a GraphQL query
type GraphQLLocation struct {
	Line   int `json:"line"`
	Column int `json:"column"`
}

// HandleGraphQL handles GraphQL federation requests
func (h *FederationHandler) HandleGraphQL(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	if r.Method == http.MethodGet {
		// Handle introspection query
		h.handleIntrospection(w, r)
		return
	}

	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req GraphQLRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.sendError(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	// Handle different types of GraphQL queries
	if isIntrospectionQuery(req.Query) {
		h.handleIntrospection(w, r)
		return
	}

	// Handle entity queries (federation-specific)
	if isEntityQuery(req.Query) {
		h.handleEntityQuery(w, r, req)
		return
	}

	// Handle regular context queries
	h.handleContextQuery(w, r, req)
}

// handleIntrospection returns the GraphQL schema for federation
func (h *FederationHandler) handleIntrospection(w http.ResponseWriter, r *http.Request) {
	schema := `
		extend schema @link(url: "https://specs.apollo.dev/federation/v2.3", import: ["@key", "@external", "@provides", "@requires", "@shareable"])

		type Query {
			_service: _Service!
		}

		type _Service {
			sdl: String
		}

		type ClinicalSnapshot @key(fields: "id") {
			id: ID!
			recipeId: String!
			patientId: String!
			createdAt: String!
			data: JSON
			metadata: SnapshotMetadata
		}

		type SnapshotMetadata {
			version: String!
			checksum: String!
			auditTrail: [AuditEntry!]!
			performance: PerformanceMetrics
		}

		type AuditEntry {
			timestamp: String!
			action: String!
			userId: String!
			details: JSON
		}

		type PerformanceMetrics {
			executionTimeMs: Int!
			cacheHits: Int!
			cacheMisses: Int!
			dataSourcesAccessed: [String!]!
		}

		type Recipe @key(fields: "id") {
			id: ID!
			name: String!
			version: String!
			clinicalScenario: String!
			governance: RecipeGovernance!
			dataSources: [DataSource!]!
		}

		type RecipeGovernance {
			allowLiveFetch: Boolean!
			requireAudit: Boolean!
			maxAgeSeconds: Int!
			accessControl: [String!]!
		}

		type DataSource {
			id: String!
			type: String!
			endpoint: String!
			requiredFields: [String!]!
			cacheTTL: Int!
		}

		scalar JSON

		extend type Patient @key(fields: "id") {
			id: ID! @external
			snapshots: [ClinicalSnapshot!]!
		}
	`

	response := GraphQLResponse{
		Data: map[string]interface{}{
			"_service": map[string]interface{}{
				"sdl": schema,
			},
		},
	}

	json.NewEncoder(w).Encode(response)
}

// isIntrospectionQuery checks if the query is an introspection query
func isIntrospectionQuery(query string) bool {
	return query == "query IntrospectionQuery { __schema { queryType { name } } }" ||
		   query == "{ _service { sdl } }" ||
		   query == "query { _service { sdl } }"
}

// isEntityQuery checks if the query is a federation entity query
func isEntityQuery(query string) bool {
	return query == "query($_representations:[_Any!]!){_entities(representations:$_representations){...on ClinicalSnapshot{id recipeId patientId createdAt data metadata{version checksum}}...on Recipe{id name version clinicalScenario}}}"
}

// handleEntityQuery handles federation entity resolution
func (h *FederationHandler) handleEntityQuery(w http.ResponseWriter, r *http.Request, req GraphQLRequest) {
	// Extract representations from variables
	representations, ok := req.Variables["_representations"].([]interface{})
	if !ok {
		h.sendError(w, "Invalid representations", http.StatusBadRequest)
		return
	}

	var entities []interface{}

	for _, repr := range representations {
		reprMap := repr.(map[string]interface{})
		typename := reprMap["__typename"].(string)

		switch typename {
		case "ClinicalSnapshot":
			entity := h.resolveSnapshotEntity(reprMap)
			if entity != nil {
				entities = append(entities, entity)
			}
		case "Recipe":
			entity := h.resolveRecipeEntity(reprMap)
			if entity != nil {
				entities = append(entities, entity)
			}
		case "Patient":
			entity := h.resolvePatientSnapshots(reprMap)
			if entity != nil {
				entities = append(entities, entity)
			}
		}
	}

	response := GraphQLResponse{
		Data: map[string]interface{}{
			"_entities": entities,
		},
	}

	json.NewEncoder(w).Encode(response)
}

// resolveSnapshotEntity resolves a ClinicalSnapshot entity
func (h *FederationHandler) resolveSnapshotEntity(repr map[string]interface{}) map[string]interface{} {
	id, ok := repr["id"].(string)
	if !ok {
		log.Printf("Invalid snapshot ID in representation")
		return nil
	}

	// Call the actual gRPC service to get the snapshot
	ctx := context.Background()
	snapshot, err := h.contextService.GetSnapshot(ctx, &pb.GetSnapshotRequest{
		SnapshotId:        id,
		RequestingService: "apollo-federation",
	})

	if err != nil {
		log.Printf("Error retrieving snapshot %s: %v", id, err)
		// Return minimal entity to prevent federation errors
		return map[string]interface{}{
			"__typename": "ClinicalSnapshot",
			"id":         id,
			"recipeId":   "",
			"patientId":  "",
			"createdAt":  "1970-01-01T00:00:00Z",
		}
	}

	// Convert the protobuf response to GraphQL format
	metadata := map[string]interface{}{
		"version":  "1.0.0",
		"checksum": snapshot.Checksum,
		"auditTrail": []map[string]interface{}{},
		"performance": map[string]interface{}{
			"executionTimeMs":       100,
			"cacheHits":            0,
			"cacheMisses":          1,
			"dataSourcesAccessed": []string{},
		},
	}

	// Convert assembly metadata if present
	if snapshot.AssemblyMetadata != nil {
		if auditData := snapshot.AssemblyMetadata.Fields["auditTrail"]; auditData != nil {
			metadata["auditTrail"] = auditData.AsInterface()
		}
		if perfData := snapshot.AssemblyMetadata.Fields["performance"]; perfData != nil {
			metadata["performance"] = perfData.AsInterface()
		}
	}

	return map[string]interface{}{
		"__typename": "ClinicalSnapshot",
		"id":         snapshot.Id,
		"recipeId":   snapshot.RecipeId,
		"patientId":  snapshot.PatientId,
		"createdAt":  snapshot.CreatedAt.AsTime().Format("2006-01-02T15:04:05Z"),
		"data":       snapshot.Data.AsMap(),
		"metadata":   metadata,
	}
}

// resolveRecipeEntity resolves a Recipe entity
func (h *FederationHandler) resolveRecipeEntity(repr map[string]interface{}) map[string]interface{} {
	id, ok := repr["id"].(string)
	if !ok {
		log.Printf("Invalid recipe ID in representation")
		return nil
	}

	// Call the actual gRPC service to get the recipe
	ctx := context.Background()
	recipe, err := h.contextService.LoadRecipe(ctx, &pb.LoadRecipeRequest{
		RecipeId: id,
	})

	if err != nil {
		log.Printf("Error loading recipe %s: %v", id, err)
		// Return minimal entity to prevent federation errors
		return map[string]interface{}{
			"__typename":       "Recipe",
			"id":              id,
			"name":            "Unknown Recipe",
			"version":         "0.0.0",
			"clinicalScenario": "unknown",
		}
	}

	// Convert protobuf recipe to GraphQL format
	governance := map[string]interface{}{
		"allowLiveFetch": true,
		"requireAudit":   true,
		"maxAgeSeconds":  3600,
		"accessControl":  []string{"read:patient_data"},
	}

	var dataSources []map[string]interface{}
	for _, dp := range recipe.RequiredFields {
		dataSource := map[string]interface{}{
			"id":             dp.Name,
			"type":           dp.SourceType,
			"endpoint":       "", // This would need to be resolved from data source registry
			"requiredFields": dp.Fields,
			"cacheTTL":       int(dp.MaxAgeHours * 3600), // Convert hours to seconds
		}
		dataSources = append(dataSources, dataSource)
	}

	return map[string]interface{}{
		"__typename":       "Recipe",
		"id":              recipe.RecipeId,
		"name":            recipe.RecipeName,
		"version":         recipe.Version,
		"clinicalScenario": recipe.ClinicalScenario,
		"governance":      governance,
		"dataSources":     dataSources,
	}
}

// resolvePatientSnapshots resolves snapshots for a Patient entity
func (h *FederationHandler) resolvePatientSnapshots(repr map[string]interface{}) map[string]interface{} {
	id, ok := repr["id"].(string)
	if !ok {
		return nil
	}

	// TODO: Implement actual snapshot retrieval for patient via gRPC
	return map[string]interface{}{
		"__typename": "Patient",
		"id":        id,
		"snapshots": []map[string]interface{}{
			{
				"id":        "snapshot-123",
				"recipeId":  "recipe-456",
				"patientId": id,
				"createdAt": "2024-01-01T00:00:00Z",
			},
		},
	}
}

// handleContextQuery handles regular context queries (non-federation)
func (h *FederationHandler) handleContextQuery(w http.ResponseWriter, r *http.Request, req GraphQLRequest) {
	// For now, return a simple response
	// TODO: Implement actual query resolution
	response := GraphQLResponse{
		Data: map[string]interface{}{
			"message": "Context service query handling not implemented yet",
		},
	}

	json.NewEncoder(w).Encode(response)
}

// sendError sends a GraphQL error response
func (h *FederationHandler) sendError(w http.ResponseWriter, message string, statusCode int) {
	w.WriteHeader(statusCode)
	
	response := GraphQLResponse{
		Errors: []GraphQLError{
			{
				Message: message,
			},
		},
	}

	json.NewEncoder(w).Encode(response)
}

// SetupFederationRoutes sets up federation API routes
func SetupFederationRoutes(router *mux.Router, contextService ContextGatewayService) {
	handler := NewFederationHandler(contextService)

	// Federation GraphQL endpoint
	router.HandleFunc("/api/federation", handler.HandleGraphQL).Methods("GET", "POST", "OPTIONS")

	// Add CORS headers
	router.Use(func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Access-Control-Allow-Origin", "*")
			w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
			w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization")

			if r.Method == "OPTIONS" {
				w.WriteHeader(http.StatusOK)
				return
			}

			next.ServeHTTP(w, r)
		})
	})
}

File: backend/services/clinical-data-hub-rust/src/api/federation.rs
// Federation GraphQL endpoint for Apollo Federation integration
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use warp::{Filter, Reply};

#[derive(Debug, Deserialize)]
pub struct GraphQLRequest {
    pub query: String,
    pub variables: Option<HashMap<String, serde_json::Value>>,
    #[serde(rename = "operationName")]
    pub operation_name: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct GraphQLResponse {
    pub data: Option<serde_json::Value>,
    pub errors: Option<Vec<GraphQLError>>,
}

#[derive(Debug, Serialize)]
pub struct GraphQLError {
    pub message: String,
    pub locations: Option<Vec<GraphQLLocation>>,
    pub path: Option<Vec<serde_json::Value>>,
    pub extensions: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Serialize)]
pub struct GraphQLLocation {
    pub line: i32,
    pub column: i32,
}

pub fn federation_routes() -> impl Filter<Extract = impl Reply, Error = warp::Rejection> + Clone {
    // GraphQL federation endpoint
    let graphql = warp::path!("api" / "federation")
        .and(warp::post())
        .and(warp::body::json())
        .and_then(handle_graphql_request)
        .or(warp::path!("api" / "federation")
            .and(warp::get())
            .and_then(handle_introspection));

    graphql.with(warp::cors()
        .allow_any_origin()
        .allow_headers(vec!["content-type", "authorization"])
        .allow_methods(vec!["GET", "POST", "OPTIONS"]))
}

async fn handle_graphql_request(
    req: GraphQLRequest,
) -> Result<impl Reply, warp::Rejection> {
    // Handle different types of GraphQL queries
    if is_introspection_query(&req.query) {
        return Ok(warp::reply::json(&handle_introspection_response()));
    }

    if is_entity_query(&req.query) {
        return handle_entity_query(req).await;
    }

    // Handle regular cache queries
    handle_cache_query(req).await
}

async fn handle_introspection() -> Result<impl Reply, warp::Rejection> {
    Ok(warp::reply::json(&handle_introspection_response()))
}

fn handle_introspection_response() -> GraphQLResponse {
    let schema = r#"
        type Query {
            _service: _Service!
        }

        type _Service {
            sdl: String
        }

        type CacheEntry @key(fields: "key") {
            key: String!
            value: JSON
            metadata: CacheMetadata!
            layer: CacheLayer!
        }

        type CacheMetadata {
            createdAt: String!
            expiresAt: String
            accessCount: Int!
            lastAccessed: String!
            compression: String
            size: Int!
        }

        enum CacheLayer {
            L1_MEMORY
            L2_REDIS
            L3_PERSISTENT
        }

        type CacheStats @key(fields: "layer") {
            layer: CacheLayer!
            hitRatio: Float!
            operations: CacheOperationStats!
            performance: CachePerformanceStats!
        }

        type CacheOperationStats {
            hits: Int!
            misses: Int!
            sets: Int!
            deletes: Int!
            evictions: Int!
        }

        type CachePerformanceStats {
            averageLatencyMs: Float!
            p95LatencyMs: Float!
            p99LatencyMs: Float!
            throughputOps: Float!
        }

        type DataAggregation @key(fields: "requestId") {
            requestId: String!
            sources: [String!]!
            result: JSON
            performance: AggregationPerformance!
        }

        type AggregationPerformance {
            totalTimeMs: Int!
            parallelExecutions: Int!
            cacheUtilization: Float!
            dataSourceLatencies: [SourceLatency!]!
        }

        type SourceLatency {
            source: String!
            latencyMs: Int!
            cached: Boolean!
        }

        scalar JSON

        extend type Patient @key(fields: "id") {
            id: ID! @external
            cachedData: [CacheEntry!]!
            aggregatedData: DataAggregation
        }

        extend type ClinicalSnapshot @key(fields: "id") {
            id: ID! @external
            cachePerformance: CacheStats
        }
    "#;

    GraphQLResponse {
        data: Some(serde_json::json!({
            "_service": {
                "sdl": schema
            }
        })),
        errors: None,
    }
}

fn is_introspection_query(query: &str) -> bool {
    query.contains("IntrospectionQuery")
        || query.contains("__schema")
        || query == "{ _service { sdl } }"
        || query == "query { _service { sdl } }"
}

fn is_entity_query(query: &str) -> bool {
    query.contains("_entities") && query.contains("_representations")
}

async fn handle_entity_query(req: GraphQLRequest) -> Result<impl Reply, warp::Rejection> {
    let variables = req.variables.unwrap_or_default();
    let representations = variables
        .get("_representations")
        .and_then(|v| v.as_array())
        .unwrap_or(&vec![]);

    let mut entities = Vec::new();

    for repr in representations {
        if let Some(repr_obj) = repr.as_object() {
            if let Some(typename) = repr_obj.get("__typename").and_then(|v| v.as_str()) {
                match typename {
                    "CacheEntry" => {
                        if let Some(entity) = resolve_cache_entry(repr_obj).await {
                            entities.push(entity);
                        }
                    }
                    "CacheStats" => {
                        if let Some(entity) = resolve_cache_stats(repr_obj).await {
                            entities.push(entity);
                        }
                    }
                    "DataAggregation" => {
                        if let Some(entity) = resolve_data_aggregation(repr_obj).await {
                            entities.push(entity);
                        }
                    }
                    "Patient" => {
                        if let Some(entity) = resolve_patient_cache_data(repr_obj).await {
                            entities.push(entity);
                        }
                    }
                    "ClinicalSnapshot" => {
                        if let Some(entity) = resolve_snapshot_cache_performance(repr_obj).await {
                            entities.push(entity);
                        }
                    }
                    _ => {}
                }
            }
        }
    }

    let response = GraphQLResponse {
        data: Some(serde_json::json!({
            "_entities": entities
        })),
        errors: None,
    };

    Ok(warp::reply::json(&response))
}

async fn resolve_cache_entry(repr: &serde_json::Map<String, serde_json::Value>) -> Option<serde_json::Value> {
    let key = repr.get("key")?.as_str()?;
    
    // TODO: Implement actual cache retrieval
    // For now, return mock data
    Some(serde_json::json!({
        "__typename": "CacheEntry",
        "key": key,
        "value": {"mock": "cached_data"},
        "metadata": {
            "createdAt": "2024-01-01T00:00:00Z",
            "expiresAt": "2024-01-01T01:00:00Z",
            "accessCount": 42,
            "lastAccessed": "2024-01-01T00:30:00Z",
            "compression": "zstd",
            "size": 1024
        },
        "layer": "L1_MEMORY"
    }))
}

async fn resolve_cache_stats(repr: &serde_json::Map<String, serde_json::Value>) -> Option<serde_json::Value> {
    let layer = repr.get("layer")?.as_str()?;
    
    // TODO: Implement actual cache stats retrieval
    Some(serde_json::json!({
        "__typename": "CacheStats",
        "layer": layer,
        "hitRatio": 0.95,
        "operations": {
            "hits": 9500,
            "misses": 500,
            "sets": 1000,
            "deletes": 100,
            "evictions": 50
        },
        "performance": {
            "averageLatencyMs": 0.5,
            "p95LatencyMs": 1.2,
            "p99LatencyMs": 2.1,
            "throughputOps": 10000.0
        }
    }))
}

async fn resolve_data_aggregation(repr: &serde_json::Map<String, serde_json::Value>) -> Option<serde_json::Value> {
    let request_id = repr.get("requestId")?.as_str()?;
    
    // TODO: Implement actual aggregation retrieval
    Some(serde_json::json!({
        "__typename": "DataAggregation",
        "requestId": request_id,
        "sources": ["patient_demographics", "vital_signs", "medications"],
        "result": {"aggregated": "clinical_data"},
        "performance": {
            "totalTimeMs": 85,
            "parallelExecutions": 3,
            "cacheUtilization": 0.67,
            "dataSourceLatencies": [
                {
                    "source": "patient_demographics",
                    "latencyMs": 15,
                    "cached": true
                },
                {
                    "source": "vital_signs",
                    "latencyMs": 45,
                    "cached": false
                },
                {
                    "source": "medications",
                    "latencyMs": 25,
                    "cached": true
                }
            ]
        }
    }))
}

async fn resolve_patient_cache_data(repr: &serde_json::Map<String, serde_json::Value>) -> Option<serde_json::Value> {
    let id = repr.get("id")?.as_str()?;
    
    // TODO: Implement actual patient cache data retrieval
    Some(serde_json::json!({
        "__typename": "Patient",
        "id": id,
        "cachedData": [
            {
                "key": format!("patient:{}:demographics", id),
                "value": {"name": "John Doe", "age": 45},
                "metadata": {
                    "createdAt": "2024-01-01T00:00:00Z",
                    "accessCount": 15,
                    "lastAccessed": "2024-01-01T00:45:00Z",
                    "size": 256
                },
                "layer": "L1_MEMORY"
            }
        ],
        "aggregatedData": {
            "requestId": format!("agg_{}", id),
            "sources": ["demographics", "vitals"],
            "result": {"patient_context": "complete"}
        }
    }))
}

async fn resolve_snapshot_cache_performance(repr: &serde_json::Map<String, serde_json::Value>) -> Option<serde_json::Value> {
    let id = repr.get("id")?.as_str()?;
    
    // TODO: Implement actual snapshot performance retrieval
    Some(serde_json::json!({
        "__typename": "ClinicalSnapshot",
        "id": id,
        "cachePerformance": {
            "layer": "L2_REDIS",
            "hitRatio": 0.88,
            "operations": {
                "hits": 880,
                "misses": 120,
                "sets": 100,
                "deletes": 10,
                "evictions": 5
            },
            "performance": {
                "averageLatencyMs": 3.2,
                "p95LatencyMs": 8.5,
                "p99LatencyMs": 15.2,
                "throughputOps": 2500.0
            }
        }
    }))
}

async fn handle_cache_query(req: GraphQLRequest) -> Result<impl Reply, warp::Rejection> {
    // Handle regular cache operations
    // TODO: Implement actual cache query processing
    let response = GraphQLResponse {
        data: Some(serde_json::json!({
            "message": "Clinical Data Hub cache query handling not implemented yet"
        })),
        errors: None,
    };

    Ok(warp::reply::json(&response))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_introspection_query_detection() {
        assert!(is_introspection_query("query IntrospectionQuery { __schema { queryType { name } } }"));
        assert!(is_introspection_query("{ _service { sdl } }"));
        assert!(!is_introspection_query("{ user { name } }"));
    }

    #[test]
    fn test_entity_query_detection() {
        assert!(is_entity_query("query($_representations:[_Any!]!){_entities(representations:$_representations){...on CacheEntry{key}}}"));
        assert!(!is_entity_query("{ cacheStats { hitRatio } }"));
    }
}